### ISSUE TITLE
Using hard negative examples

### ISSUE THREAD

**Issue by NathanYee** [yhenon/keras-frcnn#75]
Two questions:
1. Is it possible to include images that are background only?
2. What will happen if I annotate an object in an image as "bg"?

**Comment by LivingProgram** [comment]
@NathanYee ,
I think:
1. It is possible to include images that are background only.
2. You can annotate an object in the image as "bg", assuming you are running using the simple_parser, it is exactly the same as annotating any other class' bounding box

For reference, simple_parser.py will include the "bg" class as a separate class as shown on [line 27 of simple_parser.py](https://github.com/yhenon/keras-frcnn/blob/master/keras_frcnn/simple_parser.py#L27)

**Comment by yhenon** [comment]
The use of hard negative mining on bg regions is broken currently unfortunately. The issue is that the code tries to maintain  a balance of negative and positive samples for training the object classifier (see lines 183-217) of train_frcnn.py. So if it gets no positive regions from the RPN it will just skip training the classifier. One option would be to add a parameter to ignore this balancing - then it could check for proposals from the RPN that overlap the bg regions and train on these?

**Comment by yuayi521** [comment]
@NathanYee what's the meaning of hard negative examples?

**Comment by NathanYee** [comment]
@yuayi521 Faster RCNN learns what objects "aren't" by training on background regions. In my case, I have a lot of images that only contain the background. These are what I meant by a hard negative example as they don't contain any positive objects.

**Comment by yuayi521** [comment]
@NathanYee thanks, what are the advantages of using hard negative examples? Does it can improve the accuracy of the well-trained model?

**Comment by NathanYee** [comment]
@yuayi521 I think it might be able to improve the RPN if the hard negative examples contain background that otherwise not found in the training images. The use of hard negatives probably depends heavily on the particular data you are using. That being said, people train Faster RCNN with a positive to negative ratio of 1:1 so it might imbalance that a bit.

**Comment by sam-pochyly** [comment]
@NathanYee Are you planning on annotating the background image with random boxes that are labeled as 'bg' and passing that in with the rest of your training data?

**Comment by sam-pochyly** [comment]
@yhenon Is passing in negative examples (labeled 'bg') on an image that also contains positive examples supported? Because I would think it would still be able to balance the number of examples.
Also, the project I'm working on consists of extracting small codes from a page that contains all text so there's a lot of false positives when tested. I'm considering passing those false positives back into training labeled as 'bg'. However, its unclear to me if that will do anything since the IoU would be 0 for all those false positives. Do you have advice regarding that idea? Thanks!!!

**Comment by NathanYee** [comment]
@sam-pochyly Nope. I would be putting boxes over specific objects that I want to be considered as background. This would make those objects be used as positive examples of 'bg' when it comes to training. FRCNN will use any RPN proposal that is less than .3 IOU as a background example.

**Comment by thealchemist89** [comment]
@yuayi521, my 2 cents on hard negative examples is that they are use to prevent overfitting so that the detector can generalize well. Depends on where the detector is used & what kind of hard negative training sets you supply into the network.

**Comment by thealchemist89** [comment]
Hi @NathanYee ,

I am in the process of doing hard negative mining as I like to minimize false detections. I understand from your earlier post that you have a few background images which you used for negative mining.

Sorry if this sounds elementary, do you box the whole image as 'bg' to perform negative mining? And also, are there any parts of the code that require some tweaking?

Thank you and to all who are reading :)
